extensions/Worker.Extensions.Storage.Queues/src/QueueMessageConverter.cs

```diff
@@ -0,0 +1,130 @@
+
+using System;
+using System.Globalization;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Azure.Storage.Queues.Models;
+using Microsoft.Azure.Functions.Worker.Converters;
+using Microsoft.Azure.Functions.Worker.Core;
+using Microsoft.Azure.Functions.Worker.Extensions.Abstractions;
+using Microsoft.Azure.Functions.Worker.Storage.Queues;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.Functions.Worker
+{
+    [SupportsDeferredBinding]
+    [SupportedConverterTypes(typeof(QueueMessage), typeof(BinaryData), typeof(JObject))]
+    internal class QueueMessageConverter : IInputConverter
+    {
+        private readonly ILogger<QueueMessageConverter> _logger;
+
+        public QueueMessageConverter(ILogger<QueueMessageConverter> logger)
+        {
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public ValueTask<ConversionResult> ConvertAsync(ConverterContext context)
+        {
+            var result = context?.Source switch
+            {
+                ModelBindingData binding => ConvertFromBindingData(context, binding),
+                _ => ConversionResult.Unhandled()
+            };
+
+            return new ValueTask<ConversionResult>(result);
+        }
+
+        private ConversionResult ConvertFromBindingData(ConverterContext context, ModelBindingData modelBindingData)
+        {
+            if (!IsQueueExtension(modelBindingData))
+            {
+                return ConversionResult.Unhandled();
+            }
+
+            try
+            {
+                QueueMessage queueMessage = ExtractQueueMessageFromBindingData(modelBindingData);
+                var result = ToTargetType(context.TargetType, queueMessage);
+
+                if (result is not null)
+                {
+                    return ConversionResult.Success(result);
+                }
+            }
+            catch (Exception ex)
+            {
+                return ConversionResult.Failed(ex);
+            }
+
+            return ConversionResult.Unhandled();
+
+        }
+
+        private bool IsQueueExtension(ModelBindingData bindingData)
+        {
+            if (bindingData?.Source is not Constants.QueueExtensionName)
+            {
+                _logger.LogTrace("Source '{source}' is not supported by {converter}", bindingData?.Source, nameof(QueueMessageConverter));
+                return false;
+            }
+
+            return true;
+        }
+
+        private object? ToTargetType(Type targetType, QueueMessage queueMessage) => targetType switch
+        {
+            Type _ when targetType == typeof(QueueMessage) => queueMessage,
+            Type _ when targetType == typeof(BinaryData) => ConvertMessageContentToBinaryData(queueMessage),
+            Type _ when targetType == typeof(JObject) => ConvertMessageContentToJObject(queueMessage),
```

Comment:
`JObject` is from `Newtonsoft.Json` right? Is that intentional? I find it weird we are mixing STJ and Newtonsoft here.

Not sure if this helps/you are aware of this behavior. But STJ and Newtonsoft (and maybe other serialization libraries) have a behavior that when you deserialize as just `object`, they will deserialize to their object representation. I think that is `JsonElement` for STJ and `JObject` (or something else) for Newtonsoft. Their may be an opportunity to fit that behavior in here somehow?

Reply:
Yeah this one I'm not 100% sure about - I was trying to make sure we have parity with in-proc, and with in-proc we allow binding directly to `JObject` ([example](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Queues#binding-queue-message-to-jobject)):

```csharp
public static class QueueTriggerFunction_JObject
{
    [FunctionName("QueueTriggerFunction")]
    public static void Run(
        [QueueTrigger("sample-queue")] JObject message,
        ILogger logger)
    {
        logger.LogInformation("Received message from sample-queue, content={content}", message["content"]);
    }
}
```

So I was trying to enable the exact capability here.

I wasn't aware of the STJ/Newtonsoft behaviour you mentioned, I'm happy to go down that route to make sure we support both?

Reply:
I understand in-proc supports `JObject`, but we have moved towards STJ as the default and I don't think it is right for this extension to bring in Newtonsoft.Json. I think it should be something intentional by the customer to enable `JObject` being supported - like they intentionally switch out the `WorkerOptions.ObjectSerializer` to a Newtonsoft implementation, then JObject is supported.

This would work by having this implementation import that serializer and use it when the type is any POCO.

Reply:
I think POCOs are already supported by the extension so we don't need to handle that here, but I replaced JObject with JsonElement - let me know what you think of the new changes